why-agency/why-test-app
# Feature/raphaels cat grid

## 📋 Task

### Description of the issue/feature/bug

Implementation of the "Cat Grid Advanced" coding test.

**Link to Zero Figma**: <a href="https://www.figma.com/file/QaAUTjri39HPaSUTEoMfWQ/Cat-Grid?type=design&node-id=55-2&mode=design&t=4rUJ02sMKDhmIP1m-0">Figma</a>

**Link to Test changes**: <a href="https://why-test-app-git-feature-raphaels-cat-grid-and-why-dev.vercel.app/">Preview build on Vercel</a>

## 👩‍🏭 Solution

This PR features an implementation of a draggable and filterable grid of cat images queried from <a href="https://thecatapi.com/">The Cat API</a>.

It also comes with some chore changes, like the inclusion of <a href="https://prettier.io/">prettier</a> and the corresponding setup with eslint and tailwind.

### Implementation details:

-   Data querying is done server-side-only with next.js
-   The (unfiltered) cat grid consists of 100 cat images that are queried from <a href="https://thecatapi.com/">The Cat API</a> and distributed beyond the borders of the screen. However the layout can also handle any other amount of images if this is desired in the future.
-   Animations are base on <a href="https://www.framer.com/motion/">framer-motion</a>.
-   Filtering is implemented for four selected cat breeds. Filtering is animated with a <a href="https://www.framer.com/motion/layout-animations/">layout animation</a>. Only implemented for large devices.
-   Filtering state is saved into URL params.
-   Clicking an image shows an overlay with details for the cat breed.
-   For small devices an alternative layout is implemented. There, we can swipe only vertically and not horizontally. Also the implementation for the overlay is adjusted.

### A note on routing:

Of course, it would have been possible to implement the overlays with a nested dynamic route, e.g. `[breed_id]`.
However for the scope of this feature, I've decided against it for the following reasons:

-   The data for the cat breed details is already available on the start page (we have to query it to do the filtering logic). So we can simply pass it to the overlay component instead of querying it on a new route.
-   From a UI perspective it made sense to me to show the image of the cat that has been clicked in the overlay, not a generic cat breed image. Therefore the overlay does not depend on a fixed route.

## 🧪 Testing

No automated testing (unit or end-to-end) has been implemented in the scope of this feature.

To manually test it:

-   Got to the <a href="https://why-test-app-git-feature-raphaels-cat-grid-and-why-dev.vercel.app/">Preview build on Vercel</a>.
-   Drag around the image grid.
-   Resize the window.
-   Click some filters. Check if both url and the images update.
-   Copy paste the new url into a new window and see if it loads the filtered images.

## ☑ Checklist

-   [✅] _Have you removed all comments + console.logs?_
-   [✅] _Have you tested across browsers + devices?_
-   [✅] _Does your PR title fit with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)?_
